go-controller/pkg/node/iptables/iptables.go

```diff
@@ -19,57 +19,51 @@ type Rule struct {
 	addErrors := errors.New("")
 	var err error
 	var ipt util.IPTablesHelper
-	var exists bool
 
-	// stores valid table chains and whether they were already created or not
-	// key is ip protocol, table, chain
-	createdChains := map[iptables.Protocol]map[string]map[string]bool{
-		iptables.ProtocolIPv4: make(map[string]map[string]bool),
-		iptables.ProtocolIPv6: make(map[string]map[string]bool),
+	// stores the rules we want to program, keyed by protocol, table, chain
+	ruleMap := map[iptables.Protocol]map[string]map[string][][]string{
+		iptables.ProtocolIPv4: make(map[string]map[string][][]string),
+		iptables.ProtocolIPv6: make(map[string]map[string][][]string),
 	}
 
+	// rules can be inserted in groups if they are within the same table
+	// sort them into a proper map. Ignore rules that do not pass the filter
 	for _, r := range rules {
 		if _, ok := filter[r.Table][r.Chain]; !ok {
 			klog.V(5).Infof("Ignoring processing rule in table: %s, chain: %s with args: \"%s\" for protocol: %v ",
 				r.Table, r.Chain, strings.Join(r.Args, " "), r.Protocol)
 			continue
 		}
-		klog.V(5).Infof("Adding rule in table: %s, chain: %s with args: \"%s\" for protocol: %v ",
-			r.Table, r.Chain, strings.Join(r.Args, " "), r.Protocol)
-		if ipt, err = util.GetIPTablesHelper(r.Protocol); err != nil {
+		if _, ok := ruleMap[r.Protocol][r.Table]; !ok {
+			ruleMap[r.Protocol][r.Table] = make(map[string][][]string)
+		}
+		ruleMap[r.Protocol][r.Table][r.Chain] = append(ruleMap[r.Protocol][r.Table][r.Chain], r.Args)
+	}
+
+	// iterate ip protocol
+	for proto, tableMap := range ruleMap {
+		// get iptables helper
+		if ipt, err = util.GetIPTablesHelper(proto); err != nil {
 			addErrors = errors.Wrapf(addErrors,
-				"Failed to add iptables %s/%s rule %q: %v", r.Table, r.Chain, strings.Join(r.Args, " "), err)
+				"Failed to get iptables helper for protocol: %v, error: %v , err", proto, err)
 			continue
 		}
-		if _, ok := createdChains[r.Protocol][r.Table][r.Chain]; !ok {
-			klog.Infof("Creating table: %s chain: %s", r.Table, r.Chain)
-			if err = ipt.NewChain(r.Table, r.Chain); err != nil {
-				klog.V(5).Infof("Chain: \"%s\" in table: \"%s\" already exists, skipping creation: %v",
-					r.Chain, r.Table, err)
-			}
-			// we assume an error means it was already created
-			if _, ok := createdChains[r.Protocol][r.Table]; !ok {
-				createdChains[r.Protocol][r.Table] = make(map[string]bool)
-			}
-			createdChains[r.Protocol][r.Table][r.Chain] = true
-		}
-		exists, err = ipt.Exists(r.Table, r.Chain, r.Args...)
-		if !exists && err == nil {
-			if append {
-				err = ipt.Append(r.Table, r.Chain, r.Args...)
-			} else {
-				err = ipt.Insert(r.Table, r.Chain, 1, r.Args...)
+
+		// iterate table
+		for table, chainMap := range tableMap {
+			// config map is built for the table, configure everything now
+			err = ipt.Restore(table, chainMap)
+			if err != nil {
+				addErrors = errors.Wrapf(addErrors, "failed to add iptables group of rules for table: %s,"+
```

Comment:
nit: failed to restore so that we can distinguish this one from AddRules

Reply:
done